chasm/lib/scheduler/invoker.go

```diff
@@ -0,0 +1,218 @@
+	}
+}
+
+// recordExecuteResult updates the Invoker's internal state with the results of a
+// completed InvokerExecuteTask. Tasks to continue execution are added, if needed.
+func (i *Invoker) recordExecuteResult(ctx chasm.MutableContext, result *executeResult) (struct{}, error) {
+	completed := make(map[string]bool)                       // request ID -> is present
+	failed := make(map[string]bool)                          // request ID -> is present
+	retryable := make(map[string]*schedulespb.BufferedStart) // request ID -> *BufferedStart
+	canceled := make(map[string]bool)                        // run ID -> is present
+	terminated := make(map[string]bool)                      // run ID -> is present
+
+	for _, start := range result.CompletedStarts {
+		completed[start.RequestId] = true
+	}
+	for _, start := range result.FailedStarts {
+		failed[start.RequestId] = true
+	}
+	for _, start := range result.RetryableStarts {
+		retryable[start.RequestId] = start
+	}
+	for _, wf := range result.CompletedCancels {
+		canceled[wf.RunId] = true
+	}
+	for _, wf := range result.CompletedTerminates {
+		terminated[wf.RunId] = true
+	}
+
+	// Update Invoker state to remove completed items from their buffers.
+	i.BufferedStarts = util.FilterSlice(i.GetBufferedStarts(), func(start *schedulespb.BufferedStart) bool {
+		return !completed[start.RequestId] && !failed[start.RequestId]
+	})
+	i.CancelWorkflows = util.FilterSlice(i.GetCancelWorkflows(), func(we *commonpb.WorkflowExecution) bool {
+		return !canceled[we.RunId]
+	})
+	i.TerminateWorkflows = util.FilterSlice(i.GetTerminateWorkflows(), func(we *commonpb.WorkflowExecution) bool {
+		return !terminated[we.RunId]
+	})
+
+	// Update attempt counts and backoffs for failed/retrying starts.
+	for _, start := range i.GetBufferedStarts() {
+		if retry, ok := retryable[start.RequestId]; ok {
+			start.Attempt++
+			start.BackoffTime = retry.GetBackoffTime()
+		}
+	}
+
+	// Add tasks if other actions are backing off or still pending execution.
+	i.addTasks(ctx)
+
+	return struct{}{}, nil
+}
+
+// addTasks adds both ProcessBuffer and Execute tasks as needed. It should be
+// called when completing processing/executing tasks, to drive backoff/retry.
+func (i *Invoker) addTasks(ctx chasm.MutableContext) {
+	totalStarts := len(i.GetBufferedStarts())
+	eligibleStarts := len(i.getEligibleBufferedStarts())
+
+	// Add a ProcessBuffer pure task whenever there are BufferedStarts that are
+	// backing off, or are still pending initial processing.
+	if (totalStarts - eligibleStarts) > 0 {
+		ctx.AddTask(i, chasm.TaskAttributes{
+			ScheduledTime: i.processingDeadline(),
+		}, &schedulespb.InvokerProcessBufferTask{})
+	}
+
+	// Add an Execute side effect task whenever there are any eligible actions
+	// pending execution.
+	if len(i.CancelWorkflows) > 0 || len(i.TerminateWorkflows) > 0 || eligibleStarts > 0 {
+		ctx.AddTask(i, chasm.TaskAttributes{}, &schedulespb.InvokerExecuteTask{})
```

Comment:
Same here, what are the dedupe semantics?

Reply:
yeah make sense. I have tracked this work item.